File: TheaterSuggestions/Java/SeatsSuggestions/src/main/java/com/octo/technology/SeatsSuggestions/DeepModel/OfferSeatingPlacesAdjacentFromTheMiddleOfTheRow.java
package com.octo.technology.SeatsSuggestions.DeepModel;

import java.util.*;

import com.octo.technology.SeatsSuggestions.PricingCategory;
import com.octo.technology.SeatsSuggestions.Seat;
import com.octo.technology.SeatsSuggestions.SeatAvailability;

import java.util.stream.Collectors;

import static java.util.Comparator.comparing;

public class OfferSeatingPlacesAdjacentFromTheMiddleOfTheRow {
    final static int NO_DISTANCE = -1;

    public static List<Seat> buildAdjacentSeats(List<SeatWithDistance> seatsWithDistance, int partySize) {
        var previousSeatingPlaceWithDistance = new SeatWithDistance(getSeatingPlaceNull(), NO_DISTANCE);
        var potentialSeatsWithDistances = new ArrayList<SeatWithDistance>();
        var adjacentSeatsList = new ArrayList<AdjacentSeats>();

        var seatingPlaceSorted = seatsWithDistance.stream().sorted(comparing(s -> s.seat().number()))
                .collect(Collectors.toList());

        for (var seatingPlaceWithDistance : seatingPlaceSorted) {
            if (areAdjacent(seatingPlaceWithDistance, previousSeatingPlaceWithDistance)) {
                addSeatingPlacesToPotentials(potentialSeatsWithDistances, seatingPlaceWithDistance, previousSeatingPlaceWithDistance);
            } else {
                if (isMatching(potentialSeatsWithDistances, partySize)) {
                    CollectAdjacentSeatByDistance(partySize, potentialSeatsWithDistances, adjacentSeatsList);
                }
                potentialSeatsWithDistances = new ArrayList<>();
            }
            previousSeatingPlaceWithDistance = seatingPlaceWithDistance;
        }

        if (!potentialSeatsWithDistances.isEmpty())
        {
            CollectAdjacentSeatByDistance(partySize, potentialSeatsWithDistances, adjacentSeatsList);
        }

        if (!adjacentSeatsList.isEmpty()) {
            return computeTheBestAdjacentSeatsFromTheMiddleOfTheRow(adjacentSeatsList);
        }
        return Collections.emptyList();
    }

    private static void CollectAdjacentSeatByDistance(int partySize, ArrayList<SeatWithDistance> potentialSeatsWithDistances, ArrayList<AdjacentSeats> adjacentSeatsList) {
        var seatWithDistances = new ArrayList<SeatWithDistance>();

        for (var seatingPlaceWithDistance : potentialSeatsWithDistances.stream().sorted(comparing(SeatWithDistance::distanceFromTheMiddleOfTheRow)).collect(Collectors.toList()))
        {
            seatWithDistances.add(seatingPlaceWithDistance);

            if (seatWithDistances.size() == partySize)
            {
                adjacentSeatsList.add(new AdjacentSeats(seatWithDistances.stream().sorted(comparing(s-> s.seat().number())).collect(Collectors.toList())));
                seatWithDistances = new ArrayList<>();
            }
        }
    }

    private static Seat getSeatingPlaceNull() {
        return new Seat("#", -1, PricingCategory.Mixed, SeatAvailability.Available);
    }

    private static boolean isMatching(ArrayList<SeatWithDistance> potentialSeatsWithDistances, int partySize) {
        return potentialSeatsWithDistances.size() >= partySize;
    }

    private static void addSeatingPlacesToPotentials(ArrayList<SeatWithDistance> potentialSeatsWithDistances, SeatWithDistance seatingPlaceWithDistance, SeatWithDistance previousSeatingPlaceWithDistance) {
        if (potentialSeatsWithDistances.isEmpty()) {
            potentialSeatsWithDistances.addAll(Arrays.asList(previousSeatingPlaceWithDistance, seatingPlaceWithDistance));
        } else {
            potentialSeatsWithDistances.add(seatingPlaceWithDistance);
        }
    }

    private static boolean areAdjacent(SeatWithDistance seatingPlaceWithDistance, SeatWithDistance previousSeatingPlaceWithDistance) {
        return previousSeatingPlaceWithDistance.seat().number() + 1 == seatingPlaceWithDistance.seat().number();
    }

    private static List<Seat> computeTheBestAdjacentSeatsFromTheMiddleOfTheRow(ArrayList<AdjacentSeats> adjacentSeatsList) {
        SortedMap<Integer, List<SeatWithDistance>> adjacentSeatsNearerTheMiddle = new TreeMap<>();

        for (var adjacentSeats : adjacentSeatsList) {
            var sumOfDistanceFromTheMiddle = adjacentSeats.seatsWithDistance.stream()
                    .map(SeatWithDistance::distanceFromTheMiddleOfTheRow)
                    .reduce(Integer::sum).orElse(0);
            if (adjacentSeatsNearerTheMiddle.containsKey(sumOfDistanceFromTheMiddle)) {
                sumOfDistanceFromTheMiddle++;
            }
            adjacentSeatsNearerTheMiddle.put(sumOfDistanceFromTheMiddle, adjacentSeats.seatsWithDistance);
        }
        return adjacentSeatsNearerTheMiddle
                .get(adjacentSeatsNearerTheMiddle.firstKey()).stream()
                .map(SeatWithDistance::seat)
                .collect(Collectors.toList());
    }
}

File: TheaterSuggestions/CSharp/SeatsSuggestions/DeepModel/OfferSeatingPlacesAdjacentFromTheMiddleOfTheRow.cs
using System.Collections.Generic;
using System.Linq;

namespace SeatsSuggestions.DeepModel;

public static class OfferSeatingPlacesAdjacentFromTheMiddleOfTheRow
{
    private const int NoDistance = -1;

    public static IEnumerable<Seat> BuildAdjacentSeats(IEnumerable<SeatWithDistance> seatsWithDistance,
        SuggestionRequest suggestionRequest)
    {
        var previousSeatingPlaceWithDistance = new SeatWithDistance(new SeatNull(), NoDistance);
        var potentialSeatsWithDistances = new List<SeatWithDistance>();
        var adjacentSeatsList = new List<AdjacentSeats>();

        var seatingPlaceWithDistances = seatsWithDistance.OrderBy(s => s.Seat.Number);
        foreach (var seatingPlaceWithDistance in seatingPlaceWithDistances)
        {
            if (previousSeatingPlaceWithDistance.IsAdjacentWith(seatingPlaceWithDistance))
            {
                potentialSeatsWithDistances.Add(previousSeatingPlaceWithDistance, seatingPlaceWithDistance);
            }
            else
            {
                if (potentialSeatsWithDistances.IsMatchingThe(suggestionRequest.PartyRequested))
                    CollectAdjacentSeatByDistance(suggestionRequest.PartyRequested, potentialSeatsWithDistances,
                        adjacentSeatsList);
                potentialSeatsWithDistances = [];
            }

            previousSeatingPlaceWithDistance = seatingPlaceWithDistance;
        }

        if (potentialSeatsWithDistances.Any())
            CollectAdjacentSeatByDistance(suggestionRequest.PartyRequested, potentialSeatsWithDistances,
                adjacentSeatsList);

        if (adjacentSeatsList.Any()) return ComputeTheBest(adjacentSeatsList);

        return [];
    }

    private static IEnumerable<Seat> ComputeTheBest(List<AdjacentSeats> adjacentSeatsList)
    {
        var adjacentSeatsNearerTheMiddle = ComputeTheBestAdjacentSeatsFromTheMiddleOfTheRow(adjacentSeatsList);
        return adjacentSeatsNearerTheMiddle
            .FirstOrDefault()
            .Value
            .Select(swd => swd.Seat).ToList();
    }

    private static void CollectAdjacentSeatByDistance(int partySize,
        List<SeatWithDistance> potentialSeatsWithDistances,
        List<AdjacentSeats> adjacentSeatsList)
    {
        var seatWithDistances = new List<SeatWithDistance>();
        foreach (var seatingPlaceWithDistance in potentialSeatsWithDistances.OrderBy(s =>
                     s.DistanceFromTheMiddleOfTheRow))
        {
            seatWithDistances.Add(seatingPlaceWithDistance);
            if (seatWithDistances.Count == partySize)
            {
                adjacentSeatsList.Add(new AdjacentSeats(seatWithDistances.OrderBy(s => s.Seat.Number)));
                seatWithDistances = [];
            }
        }
    }

    private static SortedDictionary<int, List<SeatWithDistance>> ComputeTheBestAdjacentSeatsFromTheMiddleOfTheRow(
        List<AdjacentSeats> adjacentSeatsList)
    {
        var adjacentSeatsNearerTheMiddle = new SortedDictionary<int, List<SeatWithDistance>>();
        foreach (var adjacentSeats in adjacentSeatsList)
        {
            var sumOfDistanceFromTheMiddle = adjacentSeats.SeatsWithDistance
                .Select(swd => swd.DistanceFromTheMiddleOfTheRow)
                .Sum();
            // we have several AdjacentSeats with the same distance
            if (adjacentSeatsNearerTheMiddle.ContainsKey(sumOfDistanceFromTheMiddle)) sumOfDistanceFromTheMiddle++;
            adjacentSeatsNearerTheMiddle.Add(sumOfDistanceFromTheMiddle, adjacentSeats.SeatsWithDistance);
        }

        return adjacentSeatsNearerTheMiddle;
    }
}